### PR TITLE
feat: add bookmarklet

### PR DESCRIPTION
**Pull request by liuyang1520**
### Summary
- feat: add bookmarklet

To use it:
1. In browser bookmark bar, "Add Page"
1. Put any name
1. Copy the code below, update the `YOUR_API_KEY` to your API key
1. Put the code in the URL input box
1. Save
1. Open any web site
1. Click on the bookmark
1. Now Amplitude SDK with more trackings is enabled, click/change any item to track

Limitations:
- Ads blocker might block the script loading
- CSP might block the script loading, depending on the target website security configuration

The bookmarklet will:
- Load the SDK from CDN
- Initialize the SDK with your API key (and user id if updated)
- Initialize the inline [auto-tracking plugin](https://github.com/amplitude/Amplitude-TypeScript/compare/add-bookmarklet?expand=1#diff-6c206852d91cf818395e72ab4930bbffffbb605dc369e24d783b005432c9e0c2R58-R213)
  - tracking `click` events
  - tracking `change` events

Steps to generate the bookmarket:
- `cd packages/analytics-browser`, then `node ../../scripts/version/create-bookmarklet-snippet.js` (this will update the file in `packages/analytics-browser/generated/amplitude-bookmarklet-snippet.js`)
- Go back to the root of the repo, then `yarn build`
- `vim packages/analytics-browser/lib/scripts/amplitude-bookmarklet.html` to update the `YOUR_API_KEY`, `YOUR_USER_ID`, `YOUR_SERVER_ZONE`
- Create a new bookmark and copy the content to it

Final bookmarklet looks like (updated with remote loading auto-tracking plugin):
```js
javascript:!function()%7B"use strict"%3B!function(e%2Ct)%7Bvar n%3De.amplitude%7C%7C%7B_q%3A%5B%5D%2C_iq%3A%7B%7D%7D%3Bif(n.invoked)e.console%26%26console.error%26%26console.error("Amplitude snippet has been loaded.")%3Belse%7Bvar r%3Dfunction(e%2Ct)%7Be.prototype%5Bt%5D%3Dfunction()%7Breturn this._q.push(%7Bname%3At%2Cargs%3AArray.prototype.slice.call(arguments%2C0)%7D)%2Cthis%7D%7D%2Ci%3Dfunction(e%2Ct%2Cn)%7Breturn function(r)%7Be._q.push(%7Bname%3At%2Cargs%3AArray.prototype.slice.call(n%2C0)%2Cresolve%3Ar%7D)%7D%7D%2Co%3Dfunction(e%2Ct%2Cn)%7Be%5Bt%5D%3Dfunction()%7Bif(n)return%7Bpromise%3Anew Promise(i(e%2Ct%2CArray.prototype.slice.call(arguments)))%7D%7D%7D%2Cs%3Dfunction(e)%7Bfor(var t%3D0%3Bt<v.length%3Bt%2B%2B)o(e%2Cv%5Bt%5D%2C!1)%3Bfor(var n%3D0%3Bn<y.length%3Bn%2B%2B)o(e%2Cy%5Bn%5D%2C!0)%7D%3Bn.invoked%3D!0%3Bvar a%3Dt.getElementsByTagName("script")%5B0%5D%2Cu%3Dt.createElement("script")%3Bu.src%3D"https%3A%2F%2Fcdn.amplitude.com%2Flibs%2Fplugin-auto-tracking-browser-0.1.1-min.js.gz"%2Cu.async%3D!1%2Ca.parentNode.insertBefore(u%2Ca)%3Bvar c%3Dt.createElement("script")%3Bc.type%3D"text%2Fjavascript"%2Cc.integrity%3D"sha384-lI19%2FrkWkq7akQskdqbaYBssAwNImFV9Iwejq7dylnP0Yx8TyWYX1PwAoaA5xrUp"%2Cc.crossOrigin%3D"anonymous"%2Cc.async%3D!1%2Cc.src%3D"https%3A%2F%2Fcdn.amplitude.com%2Flibs%2Fanalytics-browser-2.1.3-min.js.gz"%2Cc.onload%3Dfunction()%7Be.amplitude.runQueuedFunctions%7C%7Cconsole.log("%5BAmplitude%5D Error%3A could not load SDK")%2Ce.amplitude.init("ee5ae81814cf1a43c415d9ca21c46a6"%2C"test-user"%2C%7BinstanceName%3A"amplitude-bookmarklet"%2CserverZone%3A"US"%2CingestionMetadata%3A%7BsourceName%3A"browser-typescript-bookmarklet"%2CsourceVersion%3A"1.0.0"%7D%2CoptOut%3A!1%7D)%2CamplitudeAutoTrackingPlugin%26%26amplitudeAutoTrackingPlugin.autoTrackingPlugin%26%26"function"%3D%3Dtypeof amplitudeAutoTrackingPlugin.autoTrackingPlugin%26%26e.amplitude.add(amplitudeAutoTrackingPlugin.autoTrackingPlugin())%2Calert("Amplitude is now tracking events!")%7D%2Ca.parentNode.insertBefore(c%2Ca)%3Bfor(var l%3Dfunction()%7Breturn this._q%3D%5B%5D%2Cthis%7D%2Cp%3D%5B"add"%2C"append"%2C"clearAll"%2C"prepend"%2C"set"%2C"setOnce"%2C"unset"%2C"preInsert"%2C"postInsert"%2C"remove"%2C"getUserProperties"%5D%2Cd%3D0%3Bd<p.length%3Bd%2B%2B)r(l%2Cp%5Bd%5D)%3Bn.Identify%3Dl%3Bfor(var g%3Dfunction()%7Breturn this._q%3D%5B%5D%2Cthis%7D%2Cm%3D%5B"getEventProperties"%2C"setProductId"%2C"setQuantity"%2C"setPrice"%2C"setRevenue"%2C"setRevenueType"%2C"setEventProperties"%5D%2Cf%3D0%3Bf<m.length%3Bf%2B%2B)r(g%2Cm%5Bf%5D)%3Bn.Revenue%3Dg%3Bvar v%3D%5B"getDeviceId"%2C"setDeviceId"%2C"getSessionId"%2C"setSessionId"%2C"getUserId"%2C"setUserId"%2C"setOptOut"%2C"setTransport"%2C"reset"%2C"extendSession"%5D%2Cy%3D%5B"init"%2C"add"%2C"remove"%2C"track"%2C"logEvent"%2C"identify"%2C"groupIdentify"%2C"setGroup"%2C"revenue"%2C"flush"%5D%3Bs(n)%2Cn.createInstance%3Dfunction(e)%7Breturn n._iq%5Be%5D%3D%7B_q%3A%5B%5D%7D%2Cs(n._iq%5Be%5D)%2Cn._iq%5Be%5D%7D%2Ce.amplitude%3Dn%7D%7D(window%2Cdocument)%7D()%3B%0A
```

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

File: scripts/templates/browser-bookmarklet.template.js
Comment (justin-fiedler) [inline scripts/templates/browser-bookmarklet.template.js:18]: @yuhao900914 started publishing snippets for plugins. You could maybe load this too versus including directly in this snippet via template?

Comment (liuyang1520) [reply scripts/templates/browser-bookmarklet.template.js:18]: Thanks for reviewing! Good call, we are planning to build the auto-tracking plugin as a standalone package in later phases, so it can be published following the similar pattern. That requires more review/discussion as it updates the SDK itself, this works like a temp solution. I will update this to use the published one once the plugin is done.

File: packages/analytics-browser/generated/amplitude-bookmarklet-snippet.js
Comment (liuyang1520) [reply packages/analytics-browser/generated/amplitude-bookmarklet-snippet.js]: Updated. As the final output is compressed and mangled by `terser`, the bookmarklet won't show the variable, but we should be able to replace with the real API key in Amplitude, so user doesn't need to copy then edit manually.

Comment (liuyang1520) [reply packages/analytics-browser/generated/amplitude-bookmarklet-snippet.js]: I tried several configurations with `terser`, but it doesn't seem offering the option to keep the string variable name, related links:
- https://github.com/terser/terser/issues/891
- https://github.com/terser/terser/issues/790
- https://github.com/terser/terser#compress-options

File: scripts/templates/browser-bookmarklet.template.js
Comment (yuhao900914) [reply scripts/templates/browser-bookmarklet.template.js:19]: I feel there won't require a lot of work. 
Put all things in a plugin package which also allow you to add test. All new plugins are added in the SDK v1.x branch and it's separated from the sdk logic. It's very simple to do, let me know if you need any help on this.

File: scripts/build/rollup.config.js
Comment (yuhao900914) [inline scripts/build/rollup.config.js:158]: Do we need a related iife?

File: scripts/templates/browser-bookmarklet.template.js
Comment (liuyang1520) [reply scripts/templates/browser-bookmarklet.template.js:19]: Thanks @yuhao900914 ! Sorry for the confusion, my intention was to keep this here and not touch the SDK packages until we are ready to work on the plugin and make it configurable in SDK. In later phases, we can extract this part (probably need add more logic) and refactor into the standalone package, then make it available in the SDK configuration (try to avoid handling the naming, publishing, etc. right now). Plus if I make it a separate package right now, I need to handle the order of [remote loading two scripts](https://jason-ge.medium.com/dynamically-load-javascript-files-in-order-5318ac6bcc61), requiring another refactor of the current loading snippet. Let me know your concerns or suggestions!

File: scripts/build/rollup.config.js
Comment (liuyang1520) [reply scripts/build/rollup.config.js:158]: Good question, correct me if I am wrong, I don't think so as this can reuse the [`amplitude-min.js`](https://github.com/amplitude/Amplitude-TypeScript/pull/529/files#diff-eeaf80f697e8ad62106eaf310a1656a71a844d5656021fd0d1cea0c9e4ca2c53R13).

Comment (yuhao900914) [reply scripts/build/rollup.config.js:158]: I feel the iife is for generating files? Did you try to run yarn build in your local, will that update the existing `/amplitude-min.js` or `amplitude-gtm-min.js`?

Comment (liuyang1520) [reply scripts/build/rollup.config.js:158]: Yes, `yarn build` will generate the files in `packages/analytics-browser/lib/scripts`, including `amplitude-min.js` and `amplitude-gtm-min.js`. I think [this](https://github.com/amplitude/Amplitude-TypeScript/blob/4dce40d86372a6a4682d688113782939d8d01904/packages/analytics-browser/rollup.config.js#L3) already covers the it.

File: packages/analytics-browser/generated/amplitude-bookmarklet-snippet.js
Comment (justin-fiedler) [inline packages/analytics-browser/generated/amplitude-bookmarklet-snippet.js:154]: Some other potential useful info:
* Page Title
* Alt text
* Normalized location on page

**Review by adam-cohen-amplitude — APPROVED** [review]

**Comment by adam-cohen-amplitude** [comment]
@liuyang1520 I approved this but, I'm not familiar with this code base, so it might be good to wait for another reviewer to approve too.

**Review by justin-fiedler — APPROVED** [review]

**Comment by liuyang1520** [comment]
Discussed offline with @justin-fiedler , @yuhao900914. Hold this to not merge for now. Will be working on a new package to include the auto-tracking logic, so it can be reused here, the new package will be targeting the `1.x` branch.

**Comment by liuyang1520** [comment]
@justin-fiedler , @yuhao900914 , I have updated the bookmarklet logic to load and use the auto-tracking plugin package, PTAL this PR again. Thank you!

File: scripts/templates/browser-bookmarklet.template.js
Comment (yuhao900914) [inline scripts/templates/browser-bookmarklet.template.js:27]: why this need to be false?

Comment (liuyang1520) [reply scripts/templates/browser-bookmarklet.template.js:27]: This will make sure the scripts are loading in order, so when we init the SDK, it has the auto-tracking plugin ready to use. For more information, check: https://jason-ge.medium.com/dynamically-load-javascript-files-in-order-5318ac6bcc61.

**Review by justin-fiedler — APPROVED** [review]

**Comment by justin-fiedler** [comment]
>  I have updated the bookmarklet logic to load and use the auto-tracking plugin package

Awesome thanks!

**Pull request merged**
